base/parallel.cc

```diff
@@ -0,0 +1,122 @@
+/** parallel.cc                                                    -*- C++ -*-
+    Jeremy Barnes, 5 February 2016
+    Copyright (c) 2016 Datacratic Inc.  All rights reserved.
+
+    This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.
+*/
+
+#include "parallel.h"
+#include "mldb/compiler/compiler.h"
+#include "mldb/base/exc_assert.h"
+#include "thread_pool.h"
+#include <atomic>
+#include <mutex>
+
+namespace Datacratic {
+
+void parallelMap(size_t first, size_t last,
+                 const std::function<void (size_t)> & doWork,
+                 int occupancyLimit)
+{
+    ExcAssertGreaterEqual(last, first);
+    ExcAssertLess((last - first), 1ULL << 31);
+
+    std::atomic<int> has_exc(0);
+    std::atomic<size_t> index(first);
+    std::mutex exc_mutex;
+    std::exception_ptr exc;
+
+    // This creates a thread pool that runs jobs on the default thread pool
+    ThreadPool tp;
+
+    if (occupancyLimit == -1)
+        occupancyLimit = numCpus();
+    if (occupancyLimit > (last - first))
```

Comment:
Because if occupancyLimit is -1, numCpus() is 32 but (last - first) is 5, we still want only 5 jobs to run, not 32.